### PR TITLE
Feature Slide 7,8

### PR DESCRIPTION
**Pull request by trunglnt-3212**
Based on what you’ve learned, update the app to satisfy these requirements:
1. UITableView: Display list of songs
• Cell contains UIImage, and 2 UILabel
2. UINavigationController
- NavigationController title always in large title
- When tap on a cell, navigate to detail screen
- When tap on a cell, navigate to detail screen
3. User are able to control music player even when app is in background and in lock screen
4. Evidence:
- Video:
https://user-images.githubusercontent.com/114736720/196321972-e7bdfdb3-c1c1-4f51-921a-b30a448b1b17.mp4
- Image:
IMG_2957
IMG_2959

### CODE REVIEW

File: AppMusic/DetailSongViewController.swift [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+final class DetailSongViewController: UIViewController {
+    
+    @IBOutlet private weak var songNameLabel: UILabel!
+    @IBOutlet private weak var songPerformerLabel: UILabel!
+    @IBOutlet private weak var songImageView: UIImageView!
+    @IBOutlet private weak var prevButton: UIButton!
+    @IBOutlet private weak var nextButton: UIButton!
+    @IBOutlet private weak var pausePlayButton: UIButton!
+    @IBOutlet private weak var sliderButton: UISlider!
```
Comment (phonglh-2668) [inline AppMusic/DetailSongViewController.swift:20]:
```suggestion
    @IBOutlet private weak var buttonSlider: UISlider!
```

File: AppMusic/DetailSongViewController.swift [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    @IBOutlet private weak var timeLabel: UILabel!
+    
+    private var isPlay: Bool = true
+    private var indexCurrentSong: Int = 0
+    private var listSong: [Song] = []
```
Comment (phonglh-2668) [inline AppMusic/DetailSongViewController.swift:25]:
```suggestion
    private var isPlay = true
    private var indexCurrentSong = 0
    private var listSong = [Song]()
```

File: AppMusic/DetailSongViewController.swift [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        pausePlayButton.setImage(UIImage(named: "icon_pause"), for: .normal)
+        
+        playerManager.playSound(listSong[indexCurrentSong].audio)
+
+        durationTime = setUpFormatTime(playerManager.getDuration())
+        timeLabel.text = "00:00 / " + (durationTime ?? "00:00")
+        timeLabel.textColor = .white
+        
+        // Update UI in Lock Screen
+        let image = UIImage(named: listSong[indexCurrentSong].image)!
```
Comment (phonglh-2668) [inline AppMusic/DetailSongViewController.swift:81]: not force wrap

File: AppMusic/DetailSongViewController.swift [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        let artwork = MPMediaItemArtwork.init(boundsSize: image.size, requestHandler: { (size) -> UIImage in
```
Comment (phonglh-2668) [inline AppMusic/DetailSongViewController.swift:82]:
```suggestion
        let artwork = MPMediaItemArtwork(boundsSize: image.size, requestHandler: { (size) -> UIImage in
```

File: AppMusic/DetailSongViewController.swift [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+            MPMediaItemPropertyArtist: listSong[indexCurrentSong].performer,
+            MPMediaItemPropertyArtwork: artwork
+        ]
+    }
+    
+    private func configPlayerTime() {
+        sliderButton.maximumValue = Float(playerManager.getDuration())
+        timer?.invalidate()
+        timer = nil
+        timer = Timer.scheduledTimer(timeInterval: Double(Int(sliderButton.value)), target: self, selector: #selector(DetailSongViewController.updateSlider), userInfo: nil, repeats: true)
```
Comment (phonglh-2668) [inline AppMusic/DetailSongViewController.swift:96]: break line

File: AppMusic/DetailSongViewController.swift [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+        if (isPlay){
+            pausePlayButton.setImage(UIImage(named: "icon_play"), for: .normal)
+            currentTime = playerManager.getDeviceCurrentTime()
+            playerManager.pause()
+        }
+        else{
+            pausePlayButton.setImage(UIImage(named: "icon_pause"), for: .normal)
+            playerManager.play()
+        }
+        isPlay = !isPlay
```
Comment (phonglh-2668) [inline AppMusic/DetailSongViewController.swift:109]:
```suggestion
        if isPlay {
            pausePlayButton.setImage(UIImage(named: "icon_play"), for: .normal)
            currentTime = playerManager.getDeviceCurrentTime()
            playerManager.pause()
        }
        else {
            pausePlayButton.setImage(UIImage(named: "icon_pause"), for: .normal)
            playerManager.play()
        }
        isPlay.toggle()
```

File: AppMusic/DetailSongViewController.swift [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+        let secondString = secondDuration < 10 ? String("0\(secondDuration)") : String(secondDuration)
+        return minuteString + ":" + secondString
+    }
+    
+    @objc private func updateSlider() {
+        sliderButton.value = Float(playerManager.getCurrentTime())
+        timeLabel.text = setUpFormatTime(playerManager.getCurrentTime()) + " / " + (durationTime ?? "00:00")
+    }
+
+    @IBAction func prevTap(_ sender: Any) {
```
Comment (phonglh-2668) [inline AppMusic/DetailSongViewController.swift:145]:
```suggestion
    @IBAction func prevTapAction(_ sender: Any) {
```

File: AppMusic/DetailSongViewController.swift [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+        handlePrevTap()
+    }
+    
+    @IBAction func nextTap(_ sender: Any) {
```
Comment (phonglh-2668) [inline AppMusic/DetailSongViewController.swift:149]:
```suggestion
    @IBAction func nextTapAction(_ sender: Any) {
```

File: AppMusic/DetailSongViewController.swift [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+        handleNextTap()
+    }
+    
+    @IBAction func didTapPauseOrPlay(_ sender: Any) {
```
Comment (phonglh-2668) [inline AppMusic/DetailSongViewController.swift:153]:
```suggestion
    @IBAction func didTapPauseOrPlayAction(_ sender: Any) {
```

File: AppMusic/DetailSongViewController.swift [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    }
+    
+    @IBAction func didChangeSlider(_ sender: Any) {
+        playerManager.stop()
+        playerManager.setCurrentTime(TimeInterval(sliderButton.value))
+        if(isPlay){
+            playerManager.prepareToPlay()
+            playerManager.play()
+        }
+    }
```
Comment (phonglh-2668) [inline AppMusic/DetailSongViewController.swift:164]:
```suggestion
    @IBAction func didChangeSliderAction(_ sender: Any) {
        playerManager.stop()
        playerManager.setCurrentTime(TimeInterval(sliderButton.value))
        if isPlay {
            playerManager.prepareToPlay()
            playerManager.play()
        }
    }
```

File: AppMusic/DetailSongViewController.swift [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        addActionToPlayCommand()
+        addActionToPauseCommnd()
+        addActionToPreviousCommand()
+        addActionToNextCommand()
+    }
+
+    private func addActionToPlayCommand() {
+            MPRemoteCommandCenter.shared().playCommand.isEnabled = true
+            MPRemoteCommandCenter.shared().playCommand.addTarget(self, action: #selector(playOrPauseButtonTapped))
+    }
```
Comment (phonglh-2668) [inline AppMusic/DetailSongViewController.swift:179]: format

File: AppMusic/DetailSongViewController.swift [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    
+    @objc
+    private func nextButtonTapped() -> MPRemoteCommandHandlerStatus {
+        handleNextTap()
+        return .success
+    }
+}
+
+
```
Comment (phonglh-2668) [inline AppMusic/DetailSongViewController.swift:215]: remove blank

File: AppMusic/ListSongViewController.swift [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+//  AppMusic
+//
+//  Created by le.n.t.trung on 17/10/2022.
+//
+
+import UIKit
+
+final class ListSongViewController: UIViewController {
+    
+    @IBOutlet private weak var listSongTable: UITableView!
```
Comment (phonglh-2668) [inline AppMusic/ListSongViewController.swift:12]:
```suggestion
    @IBOutlet private weak var tableView: UITableView!
```

File: AppMusic/ListSongViewController.swift [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    
+    private var listSong: [Song] = []
```
Comment (phonglh-2668) [inline AppMusic/ListSongViewController.swift:14]:
```suggestion
    private var listSong = [Song]()
```

File: AppMusic/ListSongViewController.swift [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+
+}
+
+extension ListSongViewController: UITableViewDataSource, UITableViewDelegate {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return listSong.count
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "MusicInfoTableViewCell", for: indexPath) as! MusicInfoTableViewCell
```
Comment (phonglh-2668) [inline AppMusic/ListSongViewController.swift:45]: not force downcast

File: AppMusic/ListSongViewController.swift [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+        cell.bindNameSong(listSong[indexPath.row].name)
+        cell.bindPerformerSong(listSong[indexPath.row].performer)
+        cell.bindImageSong(listSong[indexPath.row].image)
+        return cell
+    }
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        let storyboard = UIStoryboard(name: "Main", bundle: nil)
+        let detailSongViewController = storyboard.instantiateViewController(withIdentifier: "DetailSongViewController") as! DetailSongViewController
```
Comment (phonglh-2668) [inline AppMusic/ListSongViewController.swift:54]: above

File: AppMusic/Model/MusicInfoCell/MusicInfoTableViewCell.swift [hunk 17]
```diff
@@ -0,0 +1,39 @@
+//
+//  MusicInfoTableViewCell.swift
+//  AppMusic
+//
+//  Created by le.n.t.trung on 17/10/2022.
+//
+
+import UIKit
+
+class MusicInfoTableViewCell: UITableViewCell {
```
Comment (phonglh-2668) [inline AppMusic/Model/MusicInfoCell/MusicInfoTableViewCell.swift:10]: final

File: AppMusic/Model/MusicInfoCell/MusicInfoTableViewCell.swift [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+    
+    @IBOutlet private weak var performerSong: UILabel!
+    @IBOutlet private weak var nameSong: UILabel!
+    @IBOutlet private weak var imageSong: UIImageView!
```
Comment (phonglh-2668) [inline AppMusic/Model/MusicInfoCell/MusicInfoTableViewCell.swift:14]:
```suggestion
    @IBOutlet private weak var performerSongLabel: UILabel!
    @IBOutlet private weak var nameSongLabel: UILabel!
    @IBOutlet private weak var imageSongImageView: UIImageView!
```

File: AppMusic/Model/MusicInfoCell/MusicInfoTableViewCell.swift [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        // Initialization code
+    }
+
+    override func setSelected(_ selected: Bool, animated: Bool) {
+        super.setSelected(selected, animated: animated)
+
+        // Configure the view for the selected state
+    }
```
Comment (phonglh-2668) [inline AppMusic/Model/MusicInfoCell/MusicInfoTableViewCell.swift:25]: remove

File: AppMusic/Model/MusicInfoCell/MusicInfoTableViewCell.swift [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+        imageSong.image = UIImage(named: image)
+    }
+    
+    func bindNameSong(_ name: String) {
+        nameSong.text = name
+    }
+    
+    func bindPerformerSong(_ performer: String) {
+        performerSong.text = performer
+    }
```
Comment (phonglh-2668) [inline AppMusic/Model/MusicInfoCell/MusicInfoTableViewCell.swift:37]: nên gom thành 1 bind thôi e

File: AppMusic/Model/MusicInfoCell/MusicInfoTableViewCell.swift [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+
```
Comment (phonglh-2668) [inline AppMusic/Model/MusicInfoCell/MusicInfoTableViewCell.swift:38]: remove blank

File: AppMusic/ListSongViewController.swift [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+        detailSongViewController.bindListSong(listSong)
+        detailSongViewController.bindIndexCurrentSong(indexPath.row) 
+        self.navigationController?.pushViewController(detailSongViewController, animated: true)
+    }
+    
+
```
Comment (phonglh-2668) [inline AppMusic/ListSongViewController.swift:60]: remove blank

File: AppMusic/ListSongViewController.swift [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+        listSongTable.delegate = self
+        listSongTable.dataSource = self
+    }
+    
+    private func initListSong() {
+        listSong.append(Song(name: "Tong phu", performer: "Keyo", image: "image_0_song", audio: "TongPhuSong"))
+        listSong.append(Song(name: "Biet tim dau", performer: "Tuan Hung", image: "image_1_song", audio: "BietTimDauSong"))
+        listSong.append(Song(name: "Vo tan", performer: "Trinh Thang Binh", image: "image_2_song", audio: "VoTanSong"))
+    }
+
```
Comment (phonglh-2668) [inline AppMusic/ListSongViewController.swift:36]: remove blank

**Review by tunghs-1253 — APPROVED** [review]

**Review by phonglh-2668 — APPROVED** [review]

**Pull request merged**